agcaballero/portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,64 +14,82 @@
 
 package com.google.sps.servlets;
 
-import com.google.gson.Gson;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 import java.io.IOException;
-import java.util.ArrayList;
-import java.util.Arrays;
+import java.util.Calendar;
+import java.util.Date;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import org.joda.time.DateTimeZone;
+import org.joda.time.LocalDateTime;
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
 
 /**
  * Servlet that accepts data from the comments form &
  * posts that data to the comments.html page
  */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
-  private class Comment {
-    private final String name; // stores name of commenter
-    private final String email; // stores email of commenter
-    private final String comment; // comment itself
-
-    // constructs a Comment object
-    public Comment(String name, String email, String comment) {
-      this.name = name;
-      this.email = email;
-      this.comment = comment;
-    }
-
-    // returns comment in format: "name (email): comment"
-    public String toString() {
-      return name + " (" + email + "): " + comment;
-    }
-  }
-
-  // holds all comments that have been submitted in session so far
-  private final ArrayList<Comment> comments = new ArrayList<Comment>();
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Retrieve from Datastore entities of type "Comment"
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+    PreparedQuery results = datastore.prepare(query);
     StringBuilder commentDivs = new StringBuilder();
 
-    // build a String of divs to add to page
-    for (Comment comment : comments) {
-      commentDivs.append("<div class='comment-div'><p>" + comment + "</p></div>");
+    // Build a String of divs to hold comments to add to page
+    for (Entity entity : results.asIterable()) {
+      // Retrieve info from the Entity
+      Date timestamp = (Date) entity.getProperty("timestamp");
+      String name = (String) entity.getProperty("name");
+      String email = (String) entity.getProperty("email");
+      String comment = (String) entity.getProperty("comment");
+
+      // append current div to HTML string commentDivs
+      commentDivs.append(formatComment(timestamp, name, email, comment));
```

Comment:
FYI: This is fine, though a possibly a more encapsulated approach would have been to create a constructor for a Comment that accepts an Entity. That is, to keep the Comment class, and create instances of Comments from instances of Entities.

Reply:
I agree. It would be more modular, and also more understandable. I'll make the change.

Reply:
Done